base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,398 @@
+{
+    return ConPrintf(StdOut, L"%-*.*ls", MaxWidth, MaxWidth, String);
+}
+
+// Print a string, aligned to right.
+// MaxWidth is the width for printing.
+// return FALSE if number length is longer than MaxWidth
+BOOL PrintNum(long long Number, int MaxWidth)
+{
+    // calculate the length needed to print.
+    int PrintLength = 0;
+    long long Tmp = Number;
+    do
+    {
+        Tmp /= 10;
+        PrintLength += 1;
+    } while (Tmp);
+
+    if (PrintLength > MaxWidth)
+    {
+        return FALSE;
+    }
+    ConPrintf(StdOut, L"%*lld", MaxWidth, Number);
+
+    return TRUE;
+}
+
+// Print memory size using appropriate unit, with comma-separated number, aligned to right.
+// MaxWidth is the width for printing.
+// StartingUnit is the minimum memory unit used, 0 for Byte, 1 for KB, 2 for MB ...
+// return FALSE when failed to format.
+BOOL PrintMemory(SIZE_T MemorySizeByte, int MaxWidth, int StartingUnit)
+{
+    WCHAR *MemoryUnit[] = {L"B", L"K", L"M", L"G", L"T", L"P"};
+
+    MaxWidth -= 2; // a space and a unit will occupy 2 length.
+    if (MaxWidth <= 0)
+    {
+        return FALSE;
+    }
+
+    SIZE_T MemorySize = MemorySizeByte;
+
+    for (int i = 0; i < _countof(MemoryUnit); i++)
+    {
+        if (i >= StartingUnit)
+        {
+            // calculate the length needed to print.
+            int PrintLength = 0;
+            SIZE_T Tmp = MemorySize;
+            do
+            {
+                Tmp /= 10;
+                PrintLength += 1;
+            } while (Tmp);
+
+            if (PrintLength + (PrintLength - 1) / 3 <= MaxWidth) // (PrintLength - 1) / 3 is the length comma will take.
+            {
+                // enough to hold
+
+                // print padding space
+                PrintSpace(MaxWidth - (PrintLength + (PrintLength - 1) / 3));
+
+                int Mod = 1;
+                for (int i = 0; i < PrintLength - 1; i++)
+                {
+                    Mod *= 10;
+                }
+
+                for (int i = PrintLength - 1; i >= 0; i--)
+                {
+                    ConPrintf(StdOut, L"%d", MemorySize / Mod);
+                    MemorySize %= Mod;
+                    if (i && i % 3 == 0)
+                    {
+                        ConPrintf(StdOut, L",");
+                    }
+                    Mod /= 10;
+                }
+
+                // print the unit.
+                ConPrintf(StdOut, L" %ls", MemoryUnit[i]);
+
+                return TRUE;
+            }
+        }
+        
+        MemorySize >>= 10; // MemorySize /= 1024;
+    }
+
+    // out of MemoryUnit range.
+    return FALSE;
+}
+
+BOOL EnumProcessAndPrint(BOOL bVerbose)
+{
+    // Load ntdll.dll in order to use NtQuerySystemInformation
+    HMODULE hNtDLL = LoadLibraryW(L"Ntdll.dll");
+    if (!hNtDLL)
+    {
+        ConResMsgPrintf(StdOut, 0, IDS_ENUM_FAILED);
+        return FALSE;
+    }
+
+    NT_QUERY_SYSTEM_INFORMATION PtrNtQuerySystemInformation =
```

Comment:
Do you consider this program to be able to run on any Windows NT (including of course 2000, 2003/XP, Vista+), but not on Windows 9x?
If so you could just link tasklist at compile time with ntdll, and directly use NtQuerySystemInformation() without doing that runtime delayloading.

Reply:
it's advised by msdn to load this dynamically, as it might be altered in the future.

I think stability is more important, and I don't think there's too much difference on the performance😂

Reply:
MSDN advices and information are for 3rd-party coders writing programs. Not for developers writing core utilities and the OS itself, like we do here in ReactOS.

Reply:
Do you mean by this program can't run on windows 9x ... ?

Reply:
> Do you mean by this program can't run on windows 9x ... ?

Yes this was also a question: Since I'm not sure this tasklist and the like are designed to run on Win9x (that we don't care since that tool is part of an NT system), there should be no need to also runtime-load ntdll and its api.

Reply:
now I made it linked to ```ntdll.dll``` statically. 🤣🤣 
but I still wonder under what situation should I link an application to ntdll statically .... ?